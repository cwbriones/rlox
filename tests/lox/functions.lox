// Simple function
fun hello(name) {
    print "hello " + name;
}
hello("world");

// Recursive function
fun fib(n) {
    if (n < 2) {
        return 1;
    } else {
        return fib(n - 1) + fib(n - 2);
    }
}
print fib(9);

// Closures
var total = "global";
fun incrementer(n) {
    var total = 0;
    fun inc() {
        total = total + n;
        return total;
    }
    return inc;
}

var inc = incrementer(3);
print inc();
print inc();
print inc();

print total;

// Nested functions
fun a(one) {
    fun b(two) {
        fun c(three) {
            return one + " " + two + " " + three;
        }
        return c;
    }
    return b;
}
print a("nested")("function")("calls");
