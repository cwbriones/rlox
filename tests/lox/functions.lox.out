hello world
55
3
6
9
global
nested function calls
